/social/process/how/how-v1 (#1)
holochain/how: HOW/social/process/how/how-v1.md

---

> HOW: /social/process/how

Reviewer: ```suggestion
HOW: /social/process/how/v1
```

should we just join the how string with the version?

Author: well the above creates a new directory, I was thinking that we add the version on the file name, i.e. `how-v1`.

Reviewer: i kinda like that it creates a new directory, but don't feel super strongly, whatever you prefer

Status: Draft
Created: 2024-11-28
Version: idx/1
---

## Title
Holochain Organizational Work Default Process

## Author(s)

> mattyg <matt@buildyourweb.app>, zippy <eric@holochain.org>

Reviewer: ```suggestion
zippy <eric@holochain.org>, mattyg <matt@buildyourweb.app>
```
you are definitely the lead author

### HOW Tree

Rather than using the pattern of storing and identifying standards as a long numerical list (as in RFCs or EIPs), HOWs are identified by a semantic path in a hierarchy.  This allows their relationship and types to be more readily understood.  The HOW tree is stored directly as a file hierarchy in a git repo. The HOW tree encompasses different types of standards where the type is represented as the directory structure of in the tree.  The initial structure is:

```
HOW/
├─ info/             // General guidelines or information
├─ social/
│  ├─ process/       // Changes to HOW processes or other organizational/ecosystem processes
│  │  ├─ how/        // This standard
├─ technical/
│  ├─ core/          // Changes to core Holochain protocols
│  ├─ application/   // Ecosystem and Application standards and conventions
```

### HOW Format

> HOW standards live as markdown files in this directory structure.  The format of a HOW standard consists of a markdown header and a markdown body.  The content of the header and body MUST follow a template specified by:

Reviewer: BIPs use a folder for each proposal in case the author wants to include other files for clearer organization (i.e. images, code impelementations, etc.)

I feel like we'll run into that quickly. Maybe a proposal can either be a single markdown file or a folder containing the main markdown file at `index.md` or something?

Author: This makes sense to me, and how I was thinking of `how` as a single proposal.

1.  merging the `_requirements.md` files that live at each level of the tree including and above the leaf where the HOW is being added
2.  merging in any additional body sections or headers specified by a status change in a process flow template.

The above format specification allows the templates for parts of the HOW tree to be contextually appropriate for that part of the tree.  For example, `technical/application` standards require a "Reference Implementation" section, what would not be appropriate for a `social/process` or `info` standard.

> ### HOW Workflow

Reviewer: I would love a mermaid chart showing the state transitions in this section a la https://github.com/bitcoin/bips/blob/master/bip-0002/process.png

Reviewer: Made a basic one here (formatting is a little funky but good enough IMO): https://mermaid.live/edit#pako:eNp9Ul1rwjAU_SvhPldp-qG1D4PORuboJrTFwiiMYKOG2VRqCnPS_76abtN2cxcCuefknHMh9wSrImPgwkFSyXxONyXNU4EQynjJVpIXAgVhKhSE1CPkszUXDJ1a7Fx-6M3iXosGgzuUzOOHpkue_yJDspyT5ML0Htfd1PBXal_f9so68KL4deoFwYX9D_oa55FMY-Lf4r1gviS3SJ_MSBh-i3ujezu-Ed3J-1E9c9Uq48V9tAhIfMV1o66RG4IuUoMGOStzyrPm29VQKcgty1kKbnPNaPmWgtbi1YElPJPbM4X1plJx1tNKFtFRrMCVZcU0qPbZZXvAXdPdoUH3VIB7gndw7aHlmNYEW4Y-tmzH1uAILjaHpmVg3XZGtjV2bDyqNfgoisZAH06MkdMcEzumiQ3LVG4vimwjy6LabH-iWMZlUT61q6w2uv4EVITDDQ

We could either generate an image and include, or use the 'mermaid' code block type on github to embed the mermaid code and have it rendered.

Author: Yes, I really like that.  How had one, and we should include this.

3. Author(s) MAY get early feedback from core developers.
4. This stage is formally complete when a pull request is made against the HOW repository.  The pull request must be made using the "Create HOW draft" template. Additionally, when submitting the pull request, Author(s) MUST check "allow edits and access to secrets by maintainers".

#### Refine
This is the review stage, when Author(s) are ready for feedback from the core developers and the community at large. The status for this stage are: `REVIEW`, `LAST CALL`.  The steps in this stage are:

1. A member of the core team or requested working group will do an initial evaluation of the PR.  If the a team member of working group member is willing to take on stewardship of this HOW they will update the header status to `REVIEW` and will add the section(s) specified in `/social/process/how/_requirements.md#Process Template Section: REVIEW`.  If no requested steward is willing take on the HOW it MUST be updated to status `REJECTED` and the section(s) specified in `/social/process/how/_requirements.md#Process Template Section: REJECTED` added.
2. Community members make comments and suggested changes on the pull request.

> 3. Author(s) address feedback, making changes as desired.

Reviewer: ```suggestion
3. Author(s) address feedback, making changes as desired, and keeping all commits in the git history without squashing or force-pushing.
```

4. When most concerns appear to be addressed, the steward makes a call and changes the status to `LAST CALL` at that point the community and Author(s) have 14 days to raise and address any final concerns.

> 5. Author(s) MAY chose to set the status to `WITHDRAWN` at any time during the Refine stage, at which point the process terminates.

Reviewer: ```suggestion
5. Author(s) MAY chose to set the status to `WITHDRAWN` at any time during the Define and Refine stages, at which point the process terminates.
```

author should be able to formally withdraw a DRAFT right?

#### Align

> This is the concluding stage where either the core team or the working group decides on the HOW.   The statuses for this stage are: `REJECTED`, `ALIVE`, `DEFERRED`, `OBSOLETE`.  The steps in this stage are:

Reviewer: ```suggestion
This is the concluding stage where either the core team or the working group decides on the HOW.   The statuses for this stage are: `REJECTED`, `ALIVE`, `DEFERRED`.  The steps in this stage are:
```

I might split out OBSOLETE into another bullet point for clarity. My assumption is that is reserved for when a previously LIVE item has been superseded by a newer version. So the only path to OBSOLETE is LIVE -> OBSOLETE

1. What ever acceptance process is use by core team or working group is used (voting, consensus, coin-toss, etc) is used.
2. Given the outcome of the process the status of the HOW is updated to `REJECTED`, `DEFERRED` or `ALIVE` as appropriate along with addition of sections as per `/social/process/how/_requirements.md`

> 3. In the case of a HOW being accepted, the PR is merged into the repository.

Reviewer: ```suggestion
3. In the case of a HOW being accepted, the PR is merged into the repository, keeping the original git history without squashing.
```

### HOW Format & Templates

As described in the HOW Tree section, a HOW consists of a header and sections specified by templates that are themselves broken into parts in the tree itself so as to create a contextually coherent grammar for different types of standards.

HOW partial templates MUST be stored as a marked file named `_requirements.md`.  The format of these templates is:

- An optional header section that specifies and header items to be added to the HOW.
- A list of any sections represented in markdown as:
```
## <Section Name>
<Textual description of what MUST/SHOULD/MAY be included in the section.>
```

In the case of Process templates, there is a special format for sections to be added as part of workflow transitions. Sections to be added when status changes are made are written as:
```
## Process Template Section: <status list>
<description of what statuses this template applies to>
### <Section Name> 
<Textual description of what MUST/SHOULD/MAY be included in the section.>
... (more than one section may be specified)
```

What follows are the initial templates for the tree.  Note: As this process allows for additions to the tree these templates will be added to.

> (Copy files from tree here or just point to them?)

Reviewer: is this a "TODO" you missed?

Author: Well, it's really a question.  Should the files from the tree be in the HOW iteself, and then copied into the tree when accepted, or should the be in the tree and referred to by the HOW.  I think given that the entirety of the HOW is actually the pull request, that the latter would better thus removing duplicate text.  Thoughts?

Reviewer: Yes the latter sounds good to me
